### ISSUE TITLE
processes "rePTYed" to non-graphical VT do not survive this graphical session

### ISSUE THREAD

**Issue by jnpkrn** [systemd/systemd#15328]
**systemd version the issue has been seen with**
> 245

**Used distribution**
 > Fedora Rawhide

**Expected behaviour you didn't see**
 > In a graphical session (`sway`), run terminal (`urxvt`) and
 > something within to be sure (`ssh`), then switch to another `VT`,
 > `reptyr SAID-PROCESS`, and finally quit the graphical session
 > you've previously switched back to -- said `VT`-shifted processes
 > _shall_ survive (and hence could later be repatriated in the same
 > way when graphical session is started anew)

**Unexpected behaviour you saw**
 > said moved processes will suddenly get terminated upon shutdown
 > of the graphical session

I suspect this has something to do with cgroups and their process
tracking.  Note, however, that I didn't quit the login terminal
session once the graphical session was over, so the user on the
original `VT` remains fully active (and I have this user with
lingering enabled, just to be sure).

My use case is to preserve some terminal processes like active `ssh`
connection from within the graphical session when there's a need to
restart it fully, for instance.  I think the above procedure worked
reliably in the past, not entirely sure though.
Any idea how to satisfy this use case?

(that being said, this is not necessarily a bug, but some barrier
I am currently hitting ... any help welcome)

**Comment by arvidjaar** [comment]
> I suspect this has something to do with cgroups and their process tracking.

Most likely. `loginctl user-status` after reparenting would be interesting.

> I have this user with lingering enabled

That simply means that user systemd instance is not terminated when last session is closed. It is up to you to actually start process as user service under supervision of this instance. If you know in advance, you can use `systemd-run`.

> My use case is to preserve some terminal processes like active ssh connection from within the graphical session when there's a need to restart it fully, for instance. 

How can you guarantee that environment inherited by process matches new one? As a trivial example, new graphical session may re-start SSH agent so any reference hold by existing process becomes invalid.

**Comment by jnpkrn** [comment]
> How can you guarantee that environment inherited by process matches
> new one?  As a trivial example, new graphical session may re-start
> SSH agent so any reference hold by existing process becomes invalid.

AFAIK, the SSH agent is only required to establish the connection, not
to keep it going.  I don't think any environment propagation would
really be needed in this use case, except for what `reptyr` already
handles and maybe except for some cgroup magic that's needed here.

This is strictly limited to CLI applications (and those not attached
to DBus in particular).

* * *

Will grab `loginctl user-status` snapshots next time around, plus
will investigate how the terminate happens (who sends which signal
if possible).

**Comment by poettering** [comment]
reptyr is a horrible hack involving ptrace (i.e. it's a debugger pretty much). I am not sure we need to care about that.

This hack is never going to work cleanly, since it doesn't migrate the full user context from the old to the new session, as there#s no sane mechanism for that. To work wiht systemd's process management it would have to migrate to a different cgroup, which is the issue you are seeing.

In general: tools probe terminals for features and then cache the results. migrating them between ptys does not flush those caches, hence the tool is never going to work in the general case.

Anyway, let's close this here, it's an ugly hack, which is of course OK, but then you also have to hack around all the other implications it has (for example by migrating your process out of the originating cgroup) which is clearly outisde systemd's scope though. Hope that makes snese.

**Comment by jnpkrn** [comment]
Sorry for the noise, it eventually got explained and it wasn't as much
a `reptyr` thing as I thought.

In fact, some time back with `urxvt*` terminal , I switched to its
client-server model (`urxvt*c` and `urxvt*d`), and for the server part,
I am using a `systemd` user service (socket-based activation).  This
is where said process tracking kicks in, and anything in the process
hierarchy under will receive standard `KillMode=control-group`
treatment, propagating hence to the clients and further.

That's why I remembered things were not like this before:

> I think the above procedure worked reliably in the past

Indeed, using `urxvt*` standalone terminal window, the described
scenario works well again.

Also, the actual solution for this client-server model is to add to
the respective user unit file: `KillMode=process`.  This is the
final conclusion here (for me anyway).

P.S. The discussion was shifted towards bashing `reptyr`; well,
it works pretty reliably for my use cases -- but yeah, this has
nothing to do with `systemd`, the only overlap was around process
groups handling as imposed with using user service, something
I couldn't figure out immediately.   Sorry about that.
